SWITCHYARD-2959 - ensuring that previous version is in SY version

bfitzpat opened the pull request:
drop-down list. In this case, 2.0.0.Final

apupier commented on eclipse/plugins/org.switchyard.tools.ui/src/org/switchyard/tools/ui/common/SwitchYardSettingsGroup.java:
```diff
@@ -805,7 +805,15 @@ private void populateRuntimeVersionsList(IProgressMonitor monitor) {
+            _availableVersions.add(defaultVersion);
+        }
+
+        // for SWITCHYARD-2959 - add previous version as well 
```
I prefer not setting JIRA number, would be better to extract a method with a name like ensureDefaultAndPreviousVersionAvailable

bfitzpat replied:
Is that a stylistic thing? Because you didn't comment on it in the changes for https://github.com/jboss-switchyard/tools/pull/637

If it is a style thing and something we want to achieve going forward, I'm fine with it - but I personally like leaving breadcrumbs back to the reasons behind why certain changes were made. The goal of comments is to make the code more readable and this just gives us more information. Just my opinion of course.

bfitzpat replied:
As for popping this out to another method, I'm good with that - it holds with the principles of "Building Maintainable Software" pretty closely.

bfitzpat commented:
@apupier I have updated this PR per your suggestions. Had to do a bit of refactoring (probably more than I needed, but it made sense to move more things back to M2EUtils), but I moved the new code into a new method and added a new test.  Works locally and tests all pass. Just waiting for the PR build

apupier commented on eclipse/tests/org.switchyard.tools.ui.tests/src/org/switchyard/tools/ui/tests/CreateSwitchYardProjectTest.java:
```diff
@@ -353,4 +358,14 @@ protected void execute(IProgressMonitor monitor) throws CoreException {
+        final ArtifactVersion defaultVersion = M2EUtils.parseVersion(NewSwitchYardProjectWizard.DEFAULT_RUNTIME_VERSION);
+        final ArtifactVersion previousVersion = M2EUtils.parseVersion(NewSwitchYardProjectWizard.PREVIOUS_RUNTIME_VERSION);
+        
+        assertTrue("Versions list missing default SwitchYard Version", versions.contains(defaultVersion));
```
Optional:
I like to use AssertJ to check for collections.
For instance here, it will be something like:
assertThat(versions).as("Versions list missing default SwitchYard Version").contains(defaultVersion);
but maybe a matter of taste

bfitzpat replied:
I'm ok with the older approach, but I see your point. I'll look at more prescriptive assert statements as we go forward but don't see this as necessary for now.

apupier commented on eclipse/tests/org.switchyard.tools.ui.tests/src/org/switchyard/tools/ui/tests/CreateSwitchYardProjectTest.java:
```diff
@@ -353,4 +358,14 @@ protected void execute(IProgressMonitor monitor) throws CoreException {
         
     }
     
+    public void testVersionsListHasDefaultAndPreviousVersions() throws Exception {    
```
optional: might worth adding a second test to ensure that they are not added twice

bfitzpat replied:
As you point out below, it's a set, so they're unique.

apupier commented on eclipse/plugins/org.switchyard.tools.ui/src/org/switchyard/tools/ui/M2EUtils.java:
```diff
@@ -416,6 +420,110 @@ public static void hackIntegrationPackDependency(final Dependency dependency) {
+    public static Set<ArtifactVersion> ensureDefaultAndPreviousVersionAvailable(final Set<ArtifactVersion> versions) {
+        // add default version
+        final ArtifactVersion defaultVersion = parseVersion(NewSwitchYardProjectWizard.DEFAULT_RUNTIME_VERSION);
+        if (!versions.contains(defaultVersion)) {
```
Just realizing now: if it is a Set we don't need to check if it is contained inside first

apupier commented on eclipse/plugins/org.switchyard.tools.ui/src/org/switchyard/tools/ui/M2EUtils.java:
```diff
@@ -416,6 +420,106 @@ public static void hackIntegrationPackDependency(final Dependency dependency) {
+            final Set<ArtifactVersion> versions = filterSwitchYardVersions(resolveSwitchYardVersionRange(monitor));
+            return versions;
+        } catch (CoreException e) {
+            final Set<ArtifactVersion> versions = new LinkedHashSet<ArtifactVersion>();
```
optional: directly return new LinkedHashSet<ArtifactVersion>();

bfitzpat commented:
Thanks Aurelien
